cpp/src/join/mixed_join.cu

```diff
@@ -0,0 +1,578 @@
+  // places. However, this probably isn't worth adding any time soon since we
+  // won't be able to support AST conditions for those types anyway.
+  build_join_hash_table(build, hash_table, compare_nulls, stream);
+  auto hash_table_view = hash_table.get_device_view();
+
+  auto left_conditional_view  = table_device_view::create(left_conditional, stream);
+  auto right_conditional_view = table_device_view::create(right_conditional, stream);
+
+  // For inner joins we support optimizing the join by launching one thread for
+  // whichever table is larger rather than always using the left table.
+  detail::grid_1d const config(outer_num_rows, DEFAULT_JOIN_BLOCK_SIZE);
+  auto const shmem_size_per_block = parser.shmem_per_thread * config.num_threads_per_block;
+  join_kind const kernel_join_type =
+    join_type == join_kind::FULL_JOIN ? join_kind::LEFT_JOIN : join_type;
+
+  // If the join size data was not provided as an input, compute it here.
+  std::size_t join_size;
+  std::unique_ptr<column> matches_per_row{};
+  column_view matches_view{};
+
+  if (output_size_data.has_value()) {
+    join_size    = output_size_data->first;
+    matches_view = output_size_data->second;
+  } else {
+    // Allocate storage for the counter used to get the size of the join output
+    rmm::device_scalar<std::size_t> size(0, stream, mr);
+    CHECK_CUDA(stream.value());
+
+    matches_per_row = make_numeric_column(
+      data_type(type_to_id<size_type>()), outer_num_rows, mask_state::UNALLOCATED, stream, mr);
+    // Note that the view goes out of scope after this else statement, but the
+    // data owned by matches_per_row stays alive so the data pointer is valid.
+    matches_view                     = matches_per_row->view();
+    auto matches_mutable_view        = matches_per_row->mutable_view();
+    auto matches_per_row_mutable_ptr = matches_mutable_view.begin<size_type>();
+    if (has_nulls) {
+      compute_mixed_join_output_size<DEFAULT_JOIN_BLOCK_SIZE, true>
+        <<<config.num_blocks, config.num_threads_per_block, shmem_size_per_block, stream.value()>>>(
+          *left_conditional_view,
+          *right_conditional_view,
+          *probe_view,
+          *build_view,
+          compare_nulls,
+          kernel_join_type,
+          hash_table_view,
+          parser.device_expression_data,
+          swap_tables,
+          size.data(),
+          matches_per_row_mutable_ptr);
+    } else {
+      compute_mixed_join_output_size<DEFAULT_JOIN_BLOCK_SIZE, false>
+        <<<config.num_blocks, config.num_threads_per_block, shmem_size_per_block, stream.value()>>>(
+          *left_conditional_view,
+          *right_conditional_view,
+          *probe_view,
+          *build_view,
+          compare_nulls,
+          kernel_join_type,
+          hash_table_view,
+          parser.device_expression_data,
+          swap_tables,
+          size.data(),
+          matches_per_row_mutable_ptr);
+    }
```

Comment:
```suggestion
    compute_mixed_join_output_size<DEFAULT_JOIN_BLOCK_SIZE, has_nulls>
      <<<config.num_blocks, config.num_threads_per_block, shmem_size_per_block, stream.value()>>>(
        *left_conditional_view,
        *right_conditional_view,
        *probe_view,
        *build_view,
        compare_nulls,
        kernel_join_type,
        hash_table_view,
        parser.device_expression_data,
        swap_tables,
        size.data(),
        matches_per_row_mutable_ptr);
```

Reply:
`has_nulls` is a variable that's only known at runtime, so we have to manually dispatch to the appropriate template. Passing `has_nulls` directly as a template parameter won't compile.

Reply:
Will this work? https://godbolt.org/z/YzT56h3dW

Reply:
Nice try 😅 I don't think that's quite equivalent. The reason your example works is because despite not being explicitly declared as a compile-time constant (it's a `const bool` rather than a `constexpr bool`) `has_nulls` is in fact known at compile time. You don't even need to define `ok` in your example, you can just pass `has_nulls` directly as the template parameter. To make a fair comparison you need to define `has_nulls` according to some sort of runtime condition. For example, if you do `const bool has_nulls = num > 10;` instead it will fail.